iwyu_output.cc

```diff
@@ -1069,6 +1069,34 @@ void ProcessFullUse(OneUse* use,
   if (use->ignore_use())   // we're already ignoring it
     return;
 
+  // We normally ignore uses for builtins, but when there is a mapping defined
+  // for the symbol, we should respect that.  So, we need to determine whether
+  // the symbol has any mappings.
+  bool is_builtin_function = false;
+  if (const clang::IdentifierInfo* iden = use->decl()->getIdentifier()) {
+    is_builtin_function = iden->getBuiltinID() != 0 &&
+        !clang::Builtin::Context::isBuiltinFunc(use->symbol_name().c_str());
+  }
+
+  bool is_builtin_function_with_mappings = false;
+
+  if (is_builtin_function) {
+    for (const std::string& public_file :
```

Comment:
There's some conceptual mismatch between the variable name `is_builtin_function_with_mappings` and what this code is doing. I don't yet understand the details perfectly, but spontaneously it feels like this could be triggered even if there are not strictly any mappings involved.

I'm having a hard time getting my head around the algorithm here; could you spell out an example where the cyclic detection would trigger?

Thanks!

Reply:
I think maybe this is a variant of a more general problem/algorithm, that might as well live in `iwyu_include_picker`, but I want to make sure I understand what's going on before suggesting anything potentially stupid :)

Reply:
So, I tried moving the logic to `iwyu_include_picker` and it seemed inelegant, because it needs the proprocessor_info as well, and currently that file doesn't use any preprocessor info at all.  So instead I just pulled it out into a static function in `iwyu_output.cc`.